### PR TITLE
Fixed bugs in monthsBetweenDtgs and yearsBetweenDtgs

### PR DESCRIPTION
**Pull request by MattWindsor91**
The existing definition of yearsBetweenDtgs is sensitive to leap-year issues, as it tries to handle years as durations.  For example, the number of years between 2009-03-01 and 2008-03-01 is resolved to be 0.

This commit redefines yearsBetweenDtgs as just monthsBetweenDtgs/12. It also replaces the definition of monthsBetweenDtgs with one that is *mostly* based around naively differencing the month values, but (hopefully) does the right thing when the later date is at an earlier day of the month (or earlier time within the same day) than the earlier date.

Added quite a few tests to try to exercise as many corner cases as I could think of, but any further ideas (or counter-examples!) would be nice.

### CODE REVIEW

**Comment by sifraser** [comment]
@MattWindsor91 could we add some postconditions to these functions?

**Comment by MattWindsor91** [comment]
@sifraser good shout, I'll get on that.

I imagine two (weak-ish) postconditions would be:

- lower bound: adding the output to the earliest date always leads to a date <= the latest date
- upper bound: adding one plus the output to the earliest date always leads to a date > the latest date

I'm not sure ATM if there are any stronger postconditions that aren't overspecified?

Part of me is wondering if 'what is the minimal amount of months N do I need such that, if I iterate N+1 times adding daysInMonth(currentMonth) to my current date, I go over the target date' is a (not very nice to compute) specification of this implementation.  There's a worrying feeling in the back of my head that that's actually a subtly different idea of 'how many months between these two dates' that drifts away from 'starting at time X on the Yth of Z, how many times have we successfully gone past a hypothetical time X on the Yth of each successive month', but I can't quite summon up the brain power to figure out why.

**Comment by MattWindsor91** [comment]
@sifraser I've added postconditions (the year one is currently written in terms of months x 12 - I wasn't sure whether to try add an 'add years' function, but even if I did I suspect it'd just end up being a thin wrapper around months x 12 anyway 😁)

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt [hunk 1]
```diff
@@ -164,39 +164,42 @@ object DtgUtilities {
...
+        }
+    )
+
+    val yearsBetweenDtgs: (Dtg, Dtg) -> nat = function(
+        command = { earlierDtg, laterDtg ->
+            monthsBetweenDtgs(earlierDtg, laterDtg) / 12U
+        },
+        pre = { earlierDtg, laterDtg -> earlierDtg <= laterDtg },
+        post = { earlierDtg, laterDtg, result ->
+            val x = result.toLong()
```
Comment (sifraser) [inline kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt]: :pick: We'd want `toInteger()` here to use 'spec types'

A better fix would be to bring `addMonths/subtractMonths` in line with all the other functions and take `nat` rather than having these functions both take an int

File: kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import org.junit.runners.Parameterized
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+@RunWith(Parameterized::class)
+class TestIso8601BetweenDtgs(
+    private val name: String,
+    private val earlier: Dtg,
+    private val later: Dtg,
+    // we can't use nat here, it breaks parameterized tests
```
Comment (sifraser) [inline kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt:14]: :question: because it's a typealias? if so, wouldn't it be better to use a `ULong`?

**Review by sifraser — APPROVED** [review]
Up to you whether you want to fix `addMonths` now, but I think it certainly needs to change at some point as I think you're the third person to hit its idiosyncrasy this week

File: kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt (hunk 2 above)
Comment (MattWindsor91) [reply kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt:14]: I think (and I wish I'd made it more obvious in the comment, in hindsight) that it's more something to do with however those unsigned types are interacting with the junit glue for constructing the test class.  Even switching to `ULong` (or `nat?` etc) has the exact same issue: I just get `wrong number of arguments` on every test invocation until I switch back to native JVM primitives.

**Comment by MattWindsor91** [comment]
Gone back and forth quite a bit on the postconditions, since I realised on making `addMonths` and `subtractMonths` take `nat`s that I hadn't properly accounted for their quirky behaviour around clipping to the end of the month.  My first attempt accidentally weakened the postconditions!  I think I've managed to restate them in a way that works properly, at least until someone finds another counter-example :-)

I've also added `addYears` and `subtractYears`, and tidied up a bit of inconsistency (using `12` instead of `MonthsPerYear`, `U` instead of `u`).  Really, `addYears` and `subtractYears` need better tests than I've put in, but I suspect really we need to make parameterised test classes for these and `addMonths`/`subtractMonths`.

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt [hunk 3]
```diff
@@ -56,20 +56,76 @@ class DateProperties(private val date: Date) {
...
+                date.month + (monthsToAdd - MonthsPerYear)
+            } else {
+                date.month + monthsToAdd
+            }
+
             val nextDay = min(date.day, daysInMonth(nextYear, nextMonth))
+
             mk_Date(nextYear, nextMonth, nextDay)
         },
-        pre = { n -> (date.year * 12u + date.month).safeToInt() + n > 0 }
```
Comment (sifraser) [inline kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt:68]: :pick: this spec -- for reasons -- has a max year of 9999 so probably need the equiv of the subtract pre

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt [hunk 4]
```diff
@@ -56,20 +56,76 @@ class DateProperties(private val date: Date) {
...
+    )
+
+    val addMonths: (nat) -> Date = function(
         command = { n ->
-            val nextMonth = (date.month.safeToInt() + n - 1).mod(MonthsPerYear.safeToInt()).toNat() + 1u
-            val yearShift = (date.month.safeToInt() + n - 1).floorDiv(MonthsPerYear.safeToInt())
-            val nextYear = (date.year.safeToInt() + yearShift).toNat()
+            val yearsToAdd = n / MonthsPerYear
+            val monthsToAdd = n % MonthsPerYear
+            val willOverflow = MonthsPerYear < date.month + monthsToAdd
```
Comment (sifraser) [inline kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt]: :pick: under/overflow do make me think of going out of range rather than just the year..
```suggestion
            val willOverflowYear = MonthsPerYear < date.month + monthsToAdd
```

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt [hunk 5]
```diff
@@ -56,20 +56,76 @@ class DateProperties(private val date: Date) {
...
+
+            val nextYear = if (willOverflow) {
+                date.year + (yearsToAdd + 1u)
+            } else {
+                date.year + yearsToAdd
+            }
```
Comment (sifraser) [inline kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt]: :pick: it's probably much of a muchness here but fyi you could do 
```suggestion
            val (nextYear, nextMonth) = if (willOverflow) {
                mk_(date.year + (yearsToAdd + 1u), date.month + (monthsToAdd - MonthsPerYear))
            } else {
                mk_(date.year + yearsToAdd, date.month + monthsToAdd)
            }
```

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt [hunk 6]
```diff
@@ -162,41 +170,40 @@ object DtgUtilities {
...
+        }
+    )
+
+    val yearsBetweenDtgs = function(
+        command = { earlierDtg: Dtg, laterDtg: Dtg ->
+            monthsBetweenDtgs(earlierDtg, laterDtg) / MonthsPerYear
+        },
+        pre = { earlierDtg, laterDtg -> earlierDtg <= laterDtg },
+        post = { earlierDtg, laterDtg, result ->
+            laterDtg.functions.subtractYears(result + 1u) < earlierDtg &&
```
Comment (sifraser) [inline kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt]: :question: isn't there a scenario where this violates the precondition of subtract years?

File: kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt (hunk 2 above)
Comment (sifraser) [reply kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt:14]: TBH I forgot that the unsigned long support in Java is a bit _interesting_

File: kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+                    "leap day to leap day",
+                    earlier = midnightOn(year = 2008u, month = 2u, day = 29u),
+                    later = midnightOn(year = 2012u, month = 2u, day = 29u),
+                    months = 48,
+                    years = 4,
+                ),
+                //
+                // Other pathological cases and regressions
+                //
+                case(
```
Comment (sifraser) [inline kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt:133]: :bulb: might be worth some cases around year 0 and/or 9999

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt (hunk 6 above)
Comment (MattWindsor91) [reply kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt]: ...Yep.  😱 

Back to the drawing board again!

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt [hunk 8]
```diff
@@ -180,30 +180,55 @@ object DtgUtilities {
...
+            val inUpperBoundMonth =
+                upperBound.date.year == earlierDtg.date.year && upperBound.date.month == earlierDtg.date.month
+
+            (earlierDtg <= upperBound) and {
+                inUpperBoundMonth or {
+                    val lowerBound = laterDtg.functions.subtractMonths(result + 1u)
+                    val inLowerBoundMonth =
+                        lowerBound.date.year == earlierDtg.date.year && lowerBound.date.month == earlierDtg.date.month
+
+                    (lowerBound < earlierDtg) && inLowerBoundMonth
```
Comment (MattWindsor91) [inline kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt:204]: i'm really not sure about this postcondition - it's sound against my tests, and makes sense to me, but it's now so extremely long and convoluted that it's hard to unpack and map back to the problem at hand.

File: kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt (hunk 7 above)
Comment (MattWindsor91) [reply kazuki-toolkit/src/test/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/TestIso8601BetweenDtgs.kt:133]: Added one to guard against the postcondition failure you mentioned beforehand, and also one that has its latest date set to the latest possible date for symmetry.

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt (hunk 5 above)
Comment (MattWindsor91) [reply kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt]: I didn't realise I could do this!  That's been rolled up now, it's a lot cleaner.

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt (hunk 4 above)
Comment (MattWindsor91) [reply kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt]: True, have renamed.

File: kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt (hunk 3 above)
Comment (MattWindsor91) [reply kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt:68]: Ah I didn't notice that (I've not really noticed FirstYear/LastYear, hopefully I'm addressing the former appropriately too...)

**Comment by MattWindsor91** [comment]
Tried to address the comments from the last iteration, but the postconditions are still not my favourite.  I gave up on trying to give the year functions a fancy postcondition, so moved the previous relationship `yearsBetweenDtg = monthsBetweenDtg / 12` into the postcondition for `yearsBetweenDtg` and replaced the implementation with an analogue of the `monthsBetweenDtg` one - if this is a misstep then I'd be happy to pull it back out.
